### PR TITLE
Provide types for arrays and promises in doc comments

### PR DESCRIPTION
**Pull request by lukyanovas**

### CODE REVIEW

File: js/viz/docs/docgauges.js [hunk 1]
```diff
@@ -282,7 +282,7 @@ var baseGauge = {
             /**
             * @name basegaugeoptions_scale_majorTick_customTickValues
             * @publicName customTickValues
-            * @type array
+            * @type Array<any>
```
Comment (pavelgruba) [inline js/viz/docs/docgauges.js]: Array< number >

File: js/viz/docs/docgauges.js [hunk 2]
```diff
@@ -373,7 +373,7 @@ var baseGauge = {
             /**
             * @name basegaugeoptions_scale_minorTick_customTickValues
             * @publicName customTickValues
-            * @type array
+            * @type Array<any>
```
Comment (pavelgruba) [inline js/viz/docs/docgauges.js]: Array< number >

File: js/viz/docs/docgauges.js [hunk 3]
```diff
@@ -1157,21 +1157,21 @@ var dxBarGauge = {
     /**
     * @name dxbargaugeoptions_values
     * @publicName values
-    * @type array
+    * @type Array<any>
```
Comment (pavelgruba) [inline js/viz/docs/docgauges.js]: Array< number >

File: js/viz/docs/docgauges.js [hunk 4]
```diff
@@ -1157,21 +1157,21 @@ var dxBarGauge = {
...
     * @default []
     * @notUsedInTheme
     */
     values: [],
     /**
     * @name dxbargaugemethods_values
     * @publicName values()
-    * @return array
+    * @return Array<any>
```
Comment (pavelgruba) [inline js/viz/docs/docgauges.js]: Array< number >

File: js/viz/docs/docgauges.js [hunk 5]
```diff
@@ -1157,21 +1157,21 @@ var dxBarGauge = {
...
     */
     values: function() { },
     /**
     * @name dxbargaugemethods_values
     * @publicName values(newValues)
-    * @param1 values:array
+    * @param1 values:Array<any>
```
Comment (pavelgruba) [inline js/viz/docs/docgauges.js]: Array< number >

File: js/viz/docs/doccharts.js [hunk 6]
```diff
@@ -4111,7 +4111,7 @@ var BaseChart = {
     /**
     * @name basechartmethods_getallseries
     * @publicName getAllSeries()
-    * @return array
+    * @return Array<Object>
```
Comment (alexanderPolosatov) [inline js/viz/docs/doccharts.js]: This notation will lead to problems with usage ( for example: you can't write getAllSeries()[0].color. You should cast to 'series' type or 'any' ). I recommend you use 'any' type Instead 'Object'. In the future we will reorganize our definitions for more better usage.

**Review by pavelgruba — CHANGES_REQUESTED** [review]

File: js/viz/docs/docvectormap.js [hunk 7]
```diff
@@ -1491,21 +1491,21 @@ var dxVectorMap = {
     /**
     * @name dxvectormapmethods_viewport
     * @publicName viewport()
-    * @return array
+    * @return Array<any>
```
Comment (pavelgruba) [inline js/viz/docs/docvectormap.js]: Array< number >

**Review by pavelgruba — CHANGES_REQUESTED** [review]

**Review by alexanderPolosatov — CHANGES_REQUESTED** [review]

File: js/viz/docs/doctreemap.js [hunk 8]
```diff
@@ -67,9 +67,9 @@ var dxTreeMap = {
     * @publicName layoutAlgorithm
     * @type string | function
     * @type_function_param1 e:object
-    * @type_function_param1_field1 rect:array
+    * @type_function_param1_field1 rect:Array<any>
```
Comment (pavelgruba) [inline js/viz/docs/doctreemap.js]: Array< number >

**Review by pavelgruba — CHANGES_REQUESTED** [review]

File: js/data/data_source/data_source.js [hunk 9]
```diff
@@ -399,7 +399,7 @@ var DataSource = Class.inherit({
     /**
     * @name DataSourceMethods_items
     * @publicName items()
-    * @return array
+    * @return Array<string,object,any>
```
Comment (vconst) [inline js/data/data_source/data_source.js]: Why is not `Array<any>`?

File: js/data/query.js [hunk 10]
```diff
@@ -116,67 +116,67 @@ var query = function() {
...
 * @publicName min()
-* @return Promise
+* @return Promise<number>
 */
 /**
 * @name QueryMethods_min
 * @publicName min(getter)
 * @param1 getter:object
-* @return Promise
+* @return Promise<number>
```
Comment (vconst) [inline js/data/query.js]: Promise<number|date|string>

File: js/data/query.js [hunk 11]
```diff
@@ -116,67 +116,67 @@ var query = function() {
...
 */
 
 /**
 * @name QueryMethods_max
 * @publicName max()
-* @return Promise
+* @return Promise<number>
```
Comment (vconst) [inline js/data/query.js]: Promise<number|date|string>

File: js/data/query.js [hunk 12]
```diff
@@ -116,67 +116,67 @@ var query = function() {
...
 */
 /**
 * @name QueryMethods_max
 * @publicName max(getter)
 * @param1 getter:object
-* @return Promise
+* @return Promise<number>
```
Comment (vconst) [inline js/data/query.js]: Promise<number|date|string>

File: js/ui/data_grid/ui.data_grid.summary.js [hunk 13]
```diff
@@ -276,7 +276,7 @@ gridCore.registerModule("summary", {
                 /**
                  * @name dxDataGridOptions_summary_groupItems
                  * @publicName groupItems
-                 * @type array
+                 * @type Array<Object>
```
Comment (vconst) [inline js/ui/data_grid/ui.data_grid.summary.js:279]: Array<dxDataGridOptions_summary_groupItems>

File: js/ui/data_grid/ui.data_grid.summary.js [hunk 14]
```diff
@@ -353,7 +353,7 @@ gridCore.registerModule("summary", {
                 /**
                  * @name dxDataGridOptions_summary_totalItems
                  * @publicName totalItems
-                 * @type array
+                 * @type Array<Object>
```
Comment (vconst) [inline js/ui/data_grid/ui.data_grid.summary.js:356]: Array<dxDataGridOptions_summary_totalItems>

File: js/ui/data_grid/ui.data_grid.summary.js [hunk 15]
```diff
@@ -521,7 +521,7 @@ gridCore.registerModule("summary", {
             /**
              * @name dxDataGridOptions_sortByGroupSummaryInfo
              * @publicName sortByGroupSummaryInfo
-             * @type array
+             * @type Array<Object>
```
Comment (vconst) [inline js/ui/data_grid/ui.data_grid.summary.js:524]: Array<dxDataGridOptions_sortByGroupSummaryInfo>

File: js/ui/grid_core/ui.grid_core.data_controller.js [hunk 16]
```diff
@@ -23,7 +23,7 @@ module.exports = {
             /**
              * @name GridBaseOptions_dataSource
              * @publicName dataSource
-             * @type string|array|DataSource|DataSourceOptions
+             * @type string|Array<any>|DataSource|DataSourceOptions
```
Comment (vconst) [inline js/ui/grid_core/ui.grid_core.data_controller.js]: `Array<Object>`

File: js/ui/grid_core/ui.grid_core.data_controller.js [hunk 17]
```diff
@@ -951,7 +951,7 @@ module.exports = {
                 * @name GridBaseMethods_byKey
                 * @publicName byKey(key)
                 * @param1 key:object|string|number
-                * @return Promise
+                * @return Promise<any>
```
Comment (vconst) [inline js/ui/grid_core/ui.grid_core.data_controller.js]: `Promise<Object>`

File: js/ui/grid_core/ui.grid_core.selection.js [hunk 18]
```diff
@@ -326,7 +326,7 @@ exports.SelectionController = gridCore.Controller.inherit((function() {
         /**
          * @name GridBaseMethods_selectAll
          * @publicName selectAll()
-         * @return Promise
+         * @return Promise<any>
```
Comment (vconst) [inline js/ui/grid_core/ui.grid_core.selection.js]: `Promise<any>`

File: js/ui/grid_core/ui.grid_core.selection.js [hunk 19]
```diff
@@ -339,7 +339,7 @@ exports.SelectionController = gridCore.Controller.inherit((function() {
         /**
          * @name GridBaseMethods_deselectAll
          * @publicName deselectAll()
-         * @return Promise
+         * @return Promise<any>
```
Comment (vconst) [inline js/ui/grid_core/ui.grid_core.selection.js]: `Promise<void>`

File: js/ui/grid_core/ui.grid_core.selection.js [hunk 20]
```diff
@@ -575,17 +575,17 @@ module.exports = {
...
              * @type function(e)
              * @type_function_param1 e:object
-             * @type_function_param1_field4 currentSelectedRowKeys:array
-             * @type_function_param1_field5 currentDeselectedRowKeys:array
-             * @type_function_param1_field6 selectedRowKeys:array
-             * @type_function_param1_field7 selectedRowsData:array
+             * @type_function_param1_field4 currentSelectedRowKeys:Array<any>
+             * @type_function_param1_field5 currentDeselectedRowKeys:Array<any>
+             * @type_function_param1_field6 selectedRowKeys:Array<any>
+             * @type_function_param1_field7 selectedRowsData:Array<any>
```
Comment (vconst) [inline js/ui/grid_core/ui.grid_core.selection.js]: `Array<Object>`

File: js/ui/pivot_grid/data_source.js [hunk 21]
```diff
@@ -651,7 +651,7 @@ module.exports = Class.inherit((function() {
              * @name PivotGridDataSourceOptions_onFieldsPrepared
              * @publicName onFieldsPrepared
              * @type function(fields)
-             * @type_function_param1 fields:array
+             * @type_function_param1 fields:Array<any>
```
Comment (vconst) [inline js/ui/pivot_grid/data_source.js]: `Array<PivotGridDataSourceOptions_fields>`

File: js/ui/pivot_grid/data_source.js [hunk 22]
```diff
@@ -1457,7 +1457,7 @@ module.exports = Class.inherit((function() {
          * @name PivotGridDataSourceMethods_expandHeaderItem
          * @publicName expandHeaderItem(area, path)
          * @param1 area:string
-         * @param2 path:array
+         * @param2 path:Array<any>
```
Comment (vconst) [inline js/ui/pivot_grid/data_source.js]: `Array<Object>`

File: js/ui/pivot_grid/ui.pivot_grid.js [hunk 23]
```diff
@@ -207,7 +207,7 @@ var PivotGrid = Widget.inherit({
             /**
              * @name dxPivotGridOptions_dataSource
              * @publicName dataSource
-             * @type array|PivotGridDataSource|PivotGridDataSourceOptions
+             * @type Array<PivotGridDataSourceOptions_fields>|PivotGridDataSource|PivotGridDataSourceOptions
```
Comment (vconst) [inline js/ui/pivot_grid/ui.pivot_grid.js]: `Array<Object>`

**Review by vconst — CHANGES_REQUESTED** [review]

File: js/ui/data_grid/ui.data_grid.summary.js (hunk 13 above)
Comment (lukyanovas) [reply js/ui/data_grid/ui.data_grid.summary.js:279]: It's some kind of compromise. We should create separate objects for all these cases - but now we suppose that all nested options of array members are fields of the inner object, and declare an array as "Array<Object>"

File: js/viz/docs/doccharts.js (hunk 6 above)
Comment (lukyanovas) [reply js/viz/docs/doccharts.js]: I have changed definition to Array< baseSeriesObject >

**Review by vconst — DISMISSED** [review]

**Review by pavelgruba — DISMISSED** [review]

**Review by alexanderPolosatov — DISMISSED** [review]

**Review by DokaRus — DISMISSED** [review]

File: js/ui/tabs.js [hunk 24]
```diff
@@ -116,7 +116,7 @@ var Tabs = CollectionWidget.inherit({
             /**
             * @name dxTabsOptions_selectedItems
             * @publicName selectedItems
-            * @type array
+            * @type Array<string,Object>
```
Comment (mrjohn0011) [inline js/ui/tabs.js]: Array<string,Object,number>

File: js/ui/tag_box.js [hunk 25]
```diff
@@ -299,7 +299,7 @@ var TagBox = SelectBox.inherit({
             /**
             * @name dxTagBoxOptions_selectedItems
             * @publicName selectedItems
-            * @type array
+            * @type Array<string,Object>
```
Comment (mrjohn0011) [inline js/ui/tag_box.js]: Array<string,Object,number>

File: js/ui/tag_box.js [hunk 26]
```diff
@@ -269,7 +269,7 @@ var TagBox = SelectBox.inherit({
             /**
             * @name dxTagBoxOptions_value
             * @publicName value
-            * @type array
+            * @type Array<string,Object>
```
Comment (mrjohn0011) [inline js/ui/tag_box.js]: Array<string,Object,number>

File: js/ui/tag_box.js [hunk 27]
```diff
@@ -343,7 +343,7 @@ var TagBox = SelectBox.inherit({
              * @publicName onMultiTagPreparing
              * @extends Action
              * @type_function_param1_field4 multiTagElement:jQuery
-             * @type_function_param1_field5 selectedItems:array
+             * @type_function_param1_field5 selectedItems:Array<string,Object>
```
Comment (mrjohn0011) [inline js/ui/tag_box.js]: Array<string,Object,number>

File: js/ui/tag_box.js [hunk 28]
```diff
@@ -370,8 +370,8 @@ var TagBox = SelectBox.inherit({
             * @name dxTagBoxOptions_onSelectionChanged
             * @publicName onSelectionChanged
             * @extends Action
-            * @type_function_param1_field4 addedItems:array
-            * @type_function_param1_field5 removedItems:array
+            * @type_function_param1_field4 addedItems:Array<string,Object>
+            * @type_function_param1_field5 removedItems:Array<string,Object>
```
Comment (mrjohn0011) [inline js/ui/tag_box.js]: Array<string,Object,number>

**Review by mrjohn0011 — COMMENTED** [review]
Collection widget items can be strings, objects, and numbers. It is possible to set `items: [1, 2, 3, 4]`

File: js/framework/html/html_application.js [hunk 29]
```diff
@@ -59,7 +59,7 @@ var HtmlApplication = Application.inherit({
     /**
     * @name HtmlApplicationoptions_navigation
     * @publicName navigation
-    * @type array
+    * @type Array<Object>
```
Comment (dxbykov) [inline js/framework/html/html_application.js]: If possible, it should be `Array<dxCommand>|Array<dxCommandOptions>`

File: js/framework/html/html_application.js [hunk 30]
```diff
@@ -380,7 +380,7 @@ var HtmlApplication = Application.inherit({
     /**
      * @name HtmlApplicationfields_navigation
      * @publicName navigation
-     * @type array
+     * @type Array<Object>
```
Comment (dxbykov) [inline js/framework/html/html_application.js]: `Array<dxCommand>`

**Review by mrjohn0011 — APPROVED** [review]

**Pull request merged**